### PR TITLE
Add Endpoints to view and edit Cloud Functions in dashboard

### PR DESCRIPTION
**Pull request by dblythy**
This PR closes the oldest issue here #895, and allows developers to see their cloud endpoints in the dashboard. Also related to [81](https://github.com/parse-community/parse-dashboard/issues/81), [532](https://github.com/parse-community/parse-dashboard/issues/532), #6752.

[A small update to the Dashboard will be required.](https://github.com/parse-community/parse-dashboard/pull/1613)

This also attempts to find dependancies from the entry point, so that configurations with one cloud file that imports other files via `require()`, still show in the dashboard.

In my config, my main entry point is `main2.js`, with the content:

```
require('./functions/_User.js');
require('./functions/2FA.js');
require("./functions/Admin.js");
```

Here's how that looks in the dashboard with the PR:

Screen Shot 2020-11-02 at 2 53 48 am

And clicking on each of those links opens the file as:

Screen Shot 2020-11-02 at 2 53 57 am

Let me know what you think!

### CODE REVIEW

**Comment by mtrezza** [comment]
This is another great feature addition.

Considering that this feature involves multiple PRs / repos, could you please open an issue in which we can have a meta discussion about the approach? The benefit of meta discussions in issues rather than PRs is they usually enjoy better visibility which leads to a more engaged discussion, especially when members of existing threads are re-engaged. Alternatively, you could pick up an existing issue thread which seems most recent / appropriate.

**Comment by dblythy** [comment]
> Considering that this feature involves multiple PRs / repos, could you please open an issue in which we can have a meta discussion about the approach? 

No worries. I can do that. Would you rather a new issue or continue the discussion of the [existing issue](https://github.com/parse-community/parse-server/issues/895)? Sorry for diving into these PRs without discussing first. It pains my OCD to see 86 open issues with many being "up for grabs" for 4+ years.

**Comment by mtrezza** [comment]
> Would you rather a new issue or continue the discussion of the existing issue?

Maybe you want to use the existing issue https://github.com/parse-community/parse-server/issues/895 as it already has references to other issues and doesn't have long discussion already that for example goes about a completely different approach.

> It pains my OCD to see 86 open issues with many being "up for grabs" for 4+ years.

I can assure you that your recent PRs and feature introductions are very appreciated by the community. Practically, it is most important that there is an actual PR to discuss, which you even provide as the first step, so your ambitious approach is much appreciated. 🙂


That you already the PR ready is

**Comment by dblythy** [comment]
I have also added another endpoint that allows _writing_ cloud files, which I think is pretty cool.

The inbuilt code editor in the dashboard is pretty powerful, so I thought why not utilise it.

Although I'm pretty sure editing cloud files will still require a restart of the server? Maybe we could look at building that in.

I'm also not a react or web developer so UI might need some work.

Screen Shot 2020-11-02 at 2 11 09 pm

File: src/Routers/CloudCodeRouter.js [hunk 1]
```diff
@@ -53,6 +55,24 @@ export class CloudCodeRouter extends PromiseRouter {
...
+      middleware.promiseEnforceMasterKeyAccess,
+      CloudCodeRouter.getCloudFile
+    );
+    this.route(
+      'POST',
+      '/scripts/*',
+      middleware.promiseEnforceMasterKeyAccess,
+      CloudCodeRouter.saveCloudFile
+    );
   }
```
Comment (Moumouls) [inline src/Routers/CloudCodeRouter.js:76]: Even if master key is enforced may be we should have a Parse Server option to disable this feature 😉

**Review by Moumouls — COMMENTED** [review]
It seems that the implementation do not support multi parse server deployment, so this can result in huge inconsistency issue in prod environment ? Otherwise in local dev env the implementation looks good.

May be the way to support multi parse server deployment is to save a string version of the cloud function and then perform an equality check. Then before running the cloud function we need to perform a request to ensure that the latest version is loaded.

**Comment by dblythy** [comment]
Good point and thanks for the review @Moumouls. I’m not sure how this handles multiple servers to be honest, most of the related code was already there, I just made the endpoints available. I’ll have a look at multi-server. 

A problem with the editing feature is also that the server will need to restart for changes to take affect.

**Comment by Moumouls** [comment]
@dblythy i just pass this PR to draft since it needs some adjustments before getting everything working perfectly !

**Comment by dblythy** [comment]
@Moumouls I have added the following server parameters:

```
dashboardOptions: {
        cloudFileView: true, // fine to use on multi-instance
        cloudFileEdit: true, // local / single instance only
}
```

If this approach is ok, I will update the dashboard PR so that it only allows editing if `cloudFileEdit` is true.

**Comment by dblythy** [comment]
@mtrezza do we have any appetite for having this for the next Parse Server version (even if we strip out cloudFileEdit for now?). It would be good to be able to work it in with a dashboard release too.

**Comment by mtrezza** [comment]
Sure! I'm all for it.

**Comment by kanersan** [comment]
Looks like the 2 warnings that appear are for writing tests for the lines such as:
``` throw new Parse.Error(Parse.Error.INTERNAL_SERVER_ERROR, 'No data to save.'); ```

**Review by kanersan — APPROVED** [review]

**Comment by DiegoBM** [comment]
When should we expect this to be merged on the main branch, if there is any ETA?

**Comment by dblythy** [comment]
> When should we expect this to be merged on the main branch, if there is any ETA?

It's up to the core team 😊

**Comment by Moumouls** [comment]
@dblythy here just about the multi server support, did you add some code or some warning somewhere on the dashboard. A new developer could not understand why his app in node cluster could fall into inconsistent state (outdated code on other instances).

Also here we need an another warning/block API if parse is used into a docker image (if i remember with process.env we can detect if parse run into a Docker container), why ? because if the container is restarted and the container do not have some volume mapped ALL MODIFICATIONS will be LOST.

May be you can add some documentation about limitations on the config js, then some automatic JS doc will be generated 👍 

This is a nice feature (low code developers will love it) but we need to ensure that a developer know risks about using it.

**Comment by dblythy** [comment]
After a [discussion](https://github.com/parse-community/parse-server/issues/895#issuecomment-844003753) between @mtrezza and I, we have agreed that cloud file viewing and editing isn't necessary with Parse Server. Here's why:

Before with `parse.com`, the only way you could manage your Parse Cloud was through Parse.com's dashboard. However, as you need to run a clone of the Parse Server Example repo to use Parse Server, having the cloud file editing via the dashboard is reinventing the wheel. We want users to become familiar with industry standard IDE's, and we'd never be able to keep the dashboard's IDE to the same standard. Plus, editing cloud code on a live server without running CI checks isn't good practices.

If you have comments or suggestions for why you'd want this feature, please comment at #895. If you need any tips as to how to set up a CI/CD pipeline, please join the discussion.

**Comment by mtrezza** [comment]
> From my side i would like to say sorry may be for your precious time wasted here.

As general note, I would just like to highlight this as another example why high-level discussion in issues (rather than mixed-level discussions in PRs) are the way to go. Especially when a feature has already been developed and it seems straightforward to just submit a PR, it is in the contributors own interests to initiate a high-level discussion to receive feedback about general viability - separately from the low-level feedback about a specific implementation in a PR.

**Comment by DiegoBM** [comment]
May I ask what will be the purpose of the functions panel in the future then? Or is it meant to be removed?

Could you please describe what would be the ideal flow/process to deploy functions to Parse? Or point to a link where that is explained? Some sort of tutorial would come handy. Thank you in advance

**Comment by mtrezza** [comment]
> May I ask what will be the purpose of the functions panel in the future then? Or is it meant to be removed?

I don't think there is a "Functions" panel currently visible in the dashboard. The dashboard code still may contain artifacts of previous functionality here and there.

> Could anyone please describe what would be the ideal flow/process to deploy functions to Parse? Or point to a link where that is explained? 

"Ideal" depends on what works for you. You can do an online search for "continuous deployment" to find sophisticated examples of efficient code deployment. There is nothing specific about Parse Server when it comes to deployment pipeline. For general help with your deployment pipeline I suggest you ask for help in [DevOps SE](https://devops.stackexchange.com) or the Parse Community Forum.

**Comment by DanielRub** [comment]
I think it should work the same way as AWS lambda functions. : 

-  changes should be taken into account without reloading the server,
-  Any bug introduced in a cloud function should not affect the overall system. It should only affect the function itself. 
-  Before saving changes, some checks can be applied for the user to get some visibilty on abvious bugs.

Just Trying to put myself in the user shoes.

**Pull request closed without merging**